zebra/dplane_fpm_nl.c

```diff
@@ -0,0 +1,665 @@
+
+	return 0;
+}
+
+/**
+ * Send all RIB installed routes to the connected data plane.
+ */
+static int fpm_rib_send(struct thread *t)
+{
+	struct fpm_nl_ctx *fnc = THREAD_ARG(t);
+	rib_dest_t *dest;
+	struct route_node *rn;
+	struct route_table *rt;
+	struct zebra_dplane_ctx *ctx;
+	rib_tables_iter_t rt_iter;
+
+	rt_iter.state = RIB_TABLES_ITER_S_INIT;
+	while ((rt = rib_tables_iter_next(&rt_iter))) {
+		for (rn = route_top(rt); rn; rn = srcdest_route_next(rn)) {
+			dest = rib_dest_from_rnode(rn);
+			/* Skip bad route entries. */
+			if (dest == NULL || dest->selected_fib == NULL) {
+				continue;
+			}
+
+			/* Check for already sent routes. */
+			if (CHECK_FLAG(dest->flags, RIB_DEST_UPDATE_FPM)) {
+				continue;
+			}
+
+			/* Enqueue route install. */
+			ctx = dplane_ctx_alloc();
+			dplane_ctx_route_init(ctx, DPLANE_OP_ROUTE_INSTALL, rn,
+					      dest->selected_fib);
+			if (fpm_nl_enqueue(fnc, ctx) == -1) {
+				dplane_ctx_fini(&ctx);
+				zlog_debug("%s: buffer full, come back later",
+					   __func__);
+				thread_add_timer(zrouter.master, fpm_rib_send,
+						 fnc, 1, &fnc->t_ribwalk);
+				return 0;
+			}
+			dplane_ctx_fini(&ctx);
+
+			/* Mark as sent. */
+			SET_FLAG(dest->flags, RIB_DEST_UPDATE_FPM);
+		}
+	}
+
+	/* All RIB routes sent! */
+	fnc->rib_complete = true;
+
+	return 0;
+}
+
+/*
+ * The next three functions will handle RMAC enqueue.
+ */
+struct fpm_rmac_arg {
+	struct fpm_nl_ctx *fnc;
+	zebra_l3vni_t *zl3vni;
+};
+
+static void fpm_enqueue_rmac_table(struct hash_backet *backet, void *arg)
+{
+	struct fpm_rmac_arg *fra = arg;
+	zebra_mac_t *zrmac = backet->data;
+	struct zebra_if *zif = fra->zl3vni->vxlan_if->info;
+	const struct zebra_l2info_vxlan *vxl = &zif->l2info.vxl;
+	struct zebra_dplane_ctx *ctx;
+	struct zebra_if *br_zif;
+	vlanid_t vid;
+	bool sticky;
+
+	/* Entry already sent. */
+	if (CHECK_FLAG(zrmac->flags, ZEBRA_MAC_FPM_SENT))
+		return;
+
+	sticky = !!CHECK_FLAG(zrmac->flags,
+			      (ZEBRA_MAC_STICKY | ZEBRA_MAC_REMOTE_DEF_GW));
+	br_zif = (struct zebra_if *)(zif->brslave_info.br_if->info);
+	vid = IS_ZEBRA_IF_BRIDGE_VLAN_AWARE(br_zif) ? vxl->access_vlan : 0;
+
+	ctx = mac_update_internal(DPLANE_OP_MAC_INSTALL, fra->zl3vni->vxlan_if,
+				  zif->brslave_info.br_if, vid, &zrmac->macaddr,
+				  zrmac->fwd_info.r_vtep_ip, sticky);
+	if (fpm_nl_enqueue(fra->fnc, ctx) == -1) {
+		zlog_debug("%s: buffer full, come back later",
+			   __func__);
+		thread_add_timer(zrouter.master, fpm_rmac_send,
+				 fra->fnc, 1, &fra->fnc->t_rmacwalk);
+	}
+	dplane_ctx_fini(&ctx);
+}
+
+static void fpm_enqueue_l3vni_table(struct hash_backet *backet, void *arg)
+{
+	struct fpm_rmac_arg *fra = arg;
+	zebra_l3vni_t *zl3vni = backet->data;
+
+	fra->zl3vni = zl3vni;
+	hash_iterate(zl3vni->rmac_table, fpm_enqueue_rmac_table, zl3vni);
+}
+
+static int fpm_rmac_send(struct thread *t)
+{
+	struct fpm_rmac_arg fra;
+
+	fra.fnc = THREAD_ARG(t);
+	hash_iterate(zrouter.l3vni_table, fpm_enqueue_l3vni_table, &fra);
```

Comment:
Similar comment about support for re-using a context object in this iterative path: let's make a way to re-init a ctx so you can allocate one, stick it into the `fpm_rmac_arg`, and then re-use it during the iteration.